meta/src/ast.rs

```diff
@@ -201,10 +211,86 @@ impl Expr {
     }
 }
 
+pub struct ExprTopDownIterator {
+    current: Option<Expr>,
+    next: Option<Expr>,
+    right_branches: Vec<Expr>
```

Comment:
Of course, this makes the implementation more complicated than some quick-and-dirty DFS that you can implement recursively.

Reply:
You're not wrong :P. Took me a bit to get this working correctly.